src/electionguard/async.hpp

```diff
@@ -0,0 +1,324 @@
+#ifndef __ELECTIONGUARD_CPP_ASYNC_HPP_INCLUDED__
+#define __ELECTIONGUARD_CPP_ASYNC_HPP_INCLUDED__
+
+#include "electionguard/export.h"
+//#include "thread_pool.hpp"
+#include "log.hpp"
+
+#include <atomic>
+#include <chrono>
+#include <condition_variable>
+#include <functional>
+#include <future>
+#include <memory>
+#include <mutex>
+#include <queue>
+
+using std::atomic;
+using std::condition_variable;
+using std::future;
+using std::future_status;
+using std::mutex;
+using std::unique_lock;
+using std::unique_ptr;
+using std::vector;
+using std::chrono::seconds;
+
+namespace electionguard
+{
+    /// <summary>
+    /// A simple locking mechanism that allows the compiler to choose
+    /// how to implement the deferral.
+    ///
+    /// use like this:
+    /// AsyncSemaphore task_lock(4); //<-- count of threads
+    /// auto task = std::async(std::launch::async,
+    ///                        [](AsyncSemaphore &mutex) {
+    ///                            mutex.wait();
+    ///                            return doStuff();
+    ///                        },
+    ///                        std::ref(task_lock)));
+    /// </summary>
+    class EG_INTERNAL_API AsyncSemaphore
+    {
+      public:
+        explicit AsyncSemaphore(size_t count) : _count(count) {}
+        size_t getCount() const { return _count; };
+        void lock()
+        {
+            unique_lock<mutex> lock(_mutex);
+            _condition.wait(lock, [this] { return _count != 0; });
+            --_count;
+        }
+        void unlock()
+        {
+            unique_lock<mutex> lock(_mutex);
+            ++_count;
+            _condition.notify_one();
+        }
+
+        void wait() { std::scoped_lock _lock(*this); }
+
+      private:
+        mutex _mutex;
+        condition_variable _condition;
+        atomic<size_t> _count;
+    };
+
+    template <typename T> EG_INTERNAL_API bool isReady(const future<T> &f)
+    {
+        if (f.valid()) {
+            return f.wait_for(seconds(0)) == future_status::ready;
+        } else {
+            return false;
+        }
+    }
+
+    template <typename T>
+    EG_INTERNAL_API vector<unique_ptr<T>> wait_all(vector<future<unique_ptr<T>>> &tasks)
```

Comment:
this api surface could be generalized further but i left it like this for now since it was all i needed